<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Epidemiological models: Introduction &mdash; Pyro Tutorials 1.8.3 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/pyro.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Example: Univariate epidemiological models" href="epi_sir.html" />
    <link rel="prev" title="Example: Deep Kernel Learning" href="dkl.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html">
            <img src="_static/pyro_logo_wide.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                1.8.3
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Introductory Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="intro_long.html">Introduction to Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_rendering.html">Automatic rendering of Pyro models</a></li>
<li class="toctree-l1"><a class="reference internal" href="svi_part_i.html">SVI Part I: An Introduction to Stochastic Variational Inference in Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="svi_part_ii.html">SVI Part II: Conditional Independence, Subsampling, and Amortization</a></li>
<li class="toctree-l1"><a class="reference internal" href="svi_part_iii.html">SVI Part III: ELBO Gradient Estimators</a></li>
<li class="toctree-l1"><a class="reference internal" href="svi_part_iv.html">SVI Part IV: Tips and Tricks</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Practical Pyro and PyTorch</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="bayesian_regression.html">Bayesian Regression - Introduction (Part 1)</a></li>
<li class="toctree-l1"><a class="reference internal" href="bayesian_regression_ii.html">Bayesian Regression - Inference Algorithms (Part 2)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tensor_shapes.html">Tensor shapes in Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">Modules in Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">High-dimensional Bayesian workflow, with applications to SARS-CoV-2 strains</a></li>
<li class="toctree-l1"><a class="reference internal" href="prior_predictive.html">Interactive posterior predictives checks</a></li>
<li class="toctree-l1"><a class="reference internal" href="jit.html">Using the PyTorch JIT Compiler with Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="svi_horovod.html">Example: distributed training via Horovod</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Deep Generative Models</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="vae.html">Variational Autoencoders</a></li>
<li class="toctree-l1"><a class="reference internal" href="ss-vae.html">The Semi-Supervised VAE</a></li>
<li class="toctree-l1"><a class="reference internal" href="cvae.html">Conditional Variational Auto-encoder</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalizing_flows_i.html">Normalizing Flows - Introduction (Part 1)</a></li>
<li class="toctree-l1"><a class="reference internal" href="dmm.html">Deep Markov Model</a></li>
<li class="toctree-l1"><a class="reference internal" href="air.html">Attend Infer Repeat</a></li>
<li class="toctree-l1"><a class="reference internal" href="cevae.html">Example: Causal Effect VAE</a></li>
<li class="toctree-l1"><a class="reference internal" href="sparse_gamma.html">Example: Sparse Gamma Deep Exponential Family</a></li>
<li class="toctree-l1"><a class="reference internal" href="prodlda.html">Probabilistic Topic Modeling</a></li>
<li class="toctree-l1"><a class="reference internal" href="scanvi.html"><em>scANVI: Deep Generative Modeling for Single Cell Data with Pyro</em></a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Discrete Latent Variables</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="enumeration.html">Inference with Discrete Latent Variables</a></li>
<li class="toctree-l1"><a class="reference internal" href="gmm.html">Gaussian Mixture Model</a></li>
<li class="toctree-l1"><a class="reference internal" href="dirichlet_process_mixture.html">Dirichlet Process Mixture Models in Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="toy_mixture_model_discrete_enumeration.html">Example: Toy Mixture Model With Discrete Enumeration</a></li>
<li class="toctree-l1"><a class="reference internal" href="hmm.html">Example: Hidden Markov Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="capture_recapture.html">Example: Capture-Recapture Models (CJS Models)</a></li>
<li class="toctree-l1"><a class="reference internal" href="mixed_hmm.html">Example: hierarchical mixed-effect hidden Markov models</a></li>
<li class="toctree-l1"><a class="reference internal" href="einsum.html">Example: Discrete Factor Graph Inference with Plated Einsum</a></li>
<li class="toctree-l1"><a class="reference internal" href="lda.html">Example: Amortized Latent Dirichlet Allocation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Customizing Inference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="mle_map.html">MLE and MAP Estimation</a></li>
<li class="toctree-l1"><a class="reference internal" href="mle_map.html#Doing-the-same-thing-with-AutoGuides">Doing the same thing with AutoGuides</a></li>
<li class="toctree-l1"><a class="reference internal" href="easyguide.html">Writing guides using EasyGuide</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_objectives.html">Customizing SVI objectives and training loops</a></li>
<li class="toctree-l1"><a class="reference internal" href="boosting_bbvi.html">Boosting Black Box Variational Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="neutra.html">Example: Neural MCMC with NeuTraReparam</a></li>
<li class="toctree-l1"><a class="reference internal" href="sparse_regression.html">Example: Sparse Bayesian Linear Regression</a></li>
<li class="toctree-l1"><a class="reference internal" href="autoname_examples.html">Example: reducing boilerplate with <code class="docutils literal notranslate"><span class="pre">pyro.contrib.autoname</span></code></a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application: Time Series</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="forecasting_i.html">Forecasting I: univariate, heavy tailed</a></li>
<li class="toctree-l1"><a class="reference internal" href="forecasting_ii.html">Forecasting II: state space models</a></li>
<li class="toctree-l1"><a class="reference internal" href="forecasting_iii.html">Forecasting III: hierarchical models</a></li>
<li class="toctree-l1"><a class="reference internal" href="forecasting_dlm.html">Forecasting with Dynamic Linear Model (DLM)</a></li>
<li class="toctree-l1"><a class="reference internal" href="stable.html">Levy Stable models of Stochastic Volatility</a></li>
<li class="toctree-l1"><a class="reference internal" href="forecast_simple.html">Multivariate Forecasting</a></li>
<li class="toctree-l1"><a class="reference internal" href="timeseries.html">Example: Gaussian Process Time Series Models</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application: Gaussian Processes</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gp.html">Gaussian Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="gplvm.html">Gaussian Process Latent Variable Model</a></li>
<li class="toctree-l1"><a class="reference internal" href="bo.html">Bayesian Optimization</a></li>
<li class="toctree-l1"><a class="reference internal" href="dkl.html">Example: Deep Kernel Learning</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application: Epidemiology</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Epidemiological models: Introduction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Summary">Summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Table-of-contents">Table of contents</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Basic-workflow">Basic workflow</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Modeling">Modeling</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Generating-data">Generating data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Inference">Inference</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Prediction">Prediction</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Forecasting">Forecasting</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Advanced-modeling">Advanced modeling</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Regional-models">Regional models</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Phylogenetic-likelihoods">Phylogenetic likelihoods</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Heterogeneous-models">Heterogeneous models</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Complex-compartment-flow">Complex compartment flow</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#References">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="epi_sir.html">Example: Univariate epidemiological models</a></li>
<li class="toctree-l1"><a class="reference internal" href="epi_regional.html">Example: Regional epidemiological models</a></li>
<li class="toctree-l1"><a class="reference internal" href="sir_hmc.html">Example: Epidemiological inference via HMC</a></li>
<li class="toctree-l1"><a class="reference internal" href="logistic-growth.html">Logistic growth models of SARS-CoV-2 lineage proportions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application: Biological sequences</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="mue_profile.html">Example: Constant + MuE (Profile HMM)</a></li>
<li class="toctree-l1"><a class="reference internal" href="mue_factor.html">Example: Probabilistic PCA + MuE (FactorMuE)</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application: Experimental Design</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="working_memory.html">Designing Adaptive Experiments to Study Working Memory</a></li>
<li class="toctree-l1"><a class="reference internal" href="elections.html">Predicting the outcome of a US presidential election using Bayesian optimal experimental design</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Application: Object Tracking</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tracking_1d.html">Tracking an Unknown Number of Objects</a></li>
<li class="toctree-l1"><a class="reference internal" href="ekf.html">Kalman Filter</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other Inference Algorithms</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="baseball.html">Example: analyzing baseball stats with MCMC</a></li>
<li class="toctree-l1"><a class="reference internal" href="mcmc.html">Example: Inference with Markov Chain Monte Carlo</a></li>
<li class="toctree-l1"><a class="reference internal" href="lkj.html">Example: MCMC with an LKJ prior over covariances</a></li>
<li class="toctree-l1"><a class="reference internal" href="csis.html">Compiled Sequential Importance Sampling</a></li>
<li class="toctree-l1"><a class="reference internal" href="smcfilter.html">Example: Sequential Monte Carlo Filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="inclined_plane.html">Example: importance sampling</a></li>
<li class="toctree-l1"><a class="reference internal" href="RSA-implicature.html">The Rational Speech Act framework</a></li>
<li class="toctree-l1"><a class="reference internal" href="RSA-hyperbole.html">Understanding Hyperbole using RSA</a></li>
<li class="toctree-l1"><a class="reference internal" href="predictive_deterministic.html">Example: Utilizing Predictive and Deterministic with MCMC and SVI</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Understanding Pyro's Internals</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minipyro.html">Mini-Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="effect_handlers.html">Poutine: A Guide to Programming with Effect Handlers in Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="contrib_funsor_intro_i.html"><code class="docutils literal notranslate"><span class="pre">pyro.contrib.funsor</span></code>, a new backend for Pyro - New primitives (Part 1)</a></li>
<li class="toctree-l1"><a class="reference internal" href="contrib_funsor_intro_ii.html"><code class="docutils literal notranslate"><span class="pre">pyro.contrib.funsor</span></code>, a new backend for Pyro - Building inference algorithms (Part 2)</a></li>
<li class="toctree-l1"><a class="reference internal" href="hmm_funsor.html">Example: hidden Markov models with <code class="docutils literal notranslate"><span class="pre">pyro.contrib.funsor</span></code> and <code class="docutils literal notranslate"><span class="pre">pyroapi</span></code></a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Deprecated</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="intro_part_i.html">(DEPRECATED) An Introduction to Models in Pyro</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro_part_ii.html">(DEPRECATED) An Introduction to Inference in Pyro</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Pyro Tutorials</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Epidemiological models: Introduction</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/epi_intro.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Epidemiological-models:-Introduction">
<h1>Epidemiological models: Introduction<a class="headerlink" href="#Epidemiological-models:-Introduction" title="Permalink to this headline">¶</a></h1>
<p>This tutorial introduces the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html">pyro.contrib.epidemiology</a> module, an epidemiological modeling language with a number of black box inference algorithms. This tutorial assumes the reader is already familiar with <a class="reference external" href="http://pyro.ai/examples/intro_long.html">modeling</a>, <a class="reference external" href="http://pyro.ai/examples/intro_long.html">inference</a>, and <a class="reference external" href="http://pyro.ai/examples/tensor_shapes.html">distribution shapes</a>.</p>
<p>See also the following scripts:</p>
<ul class="simple">
<li><p><a class="reference external" href="http://pyro.ai/examples/epi_sir.html">Epidemiological models: Univariate</a></p></li>
<li><p><a class="reference external" href="http://pyro.ai/examples/epi_regional.html">Epidemiological models: Regional</a></p></li>
<li><p><a class="reference external" href="http://pyro.ai/examples/sir_hmc.html">Epidemiological inference via HMC</a></p></li>
</ul>
<section id="Summary">
<h2>Summary<a class="headerlink" href="#Summary" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>To create a new model, inherit from the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel">CompartmentalModel</a> base class.</p></li>
<li><p>Override methods <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.global_model">.global_model()</a>, <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.initialize">.initialize(params)</a>, and <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.transition">.transition(params, state, t)</a>.</p></li>
<li><p>Take care to support broadcasting and vectorized interpretation in those methods.</p></li>
<li><p>For single time series, set <code class="docutils literal notranslate"><span class="pre">population</span></code> to an integer.</p></li>
<li><p>For batched time series, let <code class="docutils literal notranslate"><span class="pre">population</span></code> be a vector, and use <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel">self.region_plate</a>.</p></li>
<li><p>For models with complex inter-compartment flows, override the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.compute_flows">.compute_flows()</a> method.</p></li>
<li><p>Flows with loops (undirected or directed) are not currently supported.</p></li>
<li><p>To perform cheap approximate inference via SVI, call the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.fit_svi">.fit_svi()</a> method.</p></li>
<li><p>To perform more expensive inference via MCMC, call the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.fit_mcmc">.fit_mcmc()</a> method.</p></li>
<li><p>To stochastically predict latent and future variables, call the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.predict">.predict()</a> method.</p></li>
</ul>
</section>
<section id="Table-of-contents">
<h2>Table of contents<a class="headerlink" href="#Table-of-contents" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="#Basic-workflow">Basic workflow</a></p>
<ul>
<li><p><a class="reference external" href="#Modeling">Modeling</a></p></li>
<li><p><a class="reference external" href="#Generating-data">Generating data</a></p></li>
<li><p><a class="reference external" href="#Inference">Inference</a></p></li>
<li><p><a class="reference external" href="#Prediction">Prediction</a></p></li>
<li><p><a class="reference external" href="#Forecasting">Forecasting</a></p></li>
</ul>
</li>
<li><p><a class="reference external" href="#Advanced-modeling">Advanced modeling</a></p>
<ul>
<li><p><a class="reference external" href="#Regional-models">Regional models</a></p></li>
<li><p><a class="reference external" href="#Phylogenetic-likelihoods">Phylogenetic likelihoods</a></p></li>
<li><p><a class="reference external" href="#Heterogeneous-models">Heterogeneous models</a></p></li>
<li><p><a class="reference external" href="#Complex-compartment-flow">Complex compartment flow</a></p></li>
</ul>
</li>
<li><p><a class="reference external" href="#References">References</a></p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">pyro</span>
<span class="kn">import</span> <span class="nn">pyro.distributions</span> <span class="k">as</span> <span class="nn">dist</span>
<span class="kn">from</span> <span class="nn">pyro.contrib.epidemiology</span> <span class="kn">import</span> <span class="n">CompartmentalModel</span><span class="p">,</span> <span class="n">binomial_dist</span><span class="p">,</span> <span class="n">infection_dist</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="k">assert</span> <span class="n">pyro</span><span class="o">.</span><span class="n">__version__</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;1.8.3&#39;</span><span class="p">)</span>
<span class="n">torch</span><span class="o">.</span><span class="n">set_default_dtype</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">double</span><span class="p">)</span>  <span class="c1"># Required for MCMC inference.</span>
<span class="n">smoke_test</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;CI&#39;</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Basic-workflow">
<h3>Basic workflow<a class="headerlink" href="#Basic-workflow" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html">pyro.contrib.epidemiology</a> module provides a modeling language for a class of stochastic discrete-time discrete-count compartmental models, together with a number of black box inference algorithms to perform joint inference on global parameters and latent variables. This modeling language is more restrictive than the full Pyro probabilistic programming language:</p>
<ul class="simple">
<li><p>control flow must be static;</p></li>
<li><p>compartmental distributions are restricted to <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.distributions.binomial_dist">binomial_dist()</a>, <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.distributions.beta_binomial_dist">beta_binomial_dist()</a>, and <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.distributions.infection_dist">infection_dist()</a>;</p></li>
<li><p>plates are not allowed, except for the single optional <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.region_plate">.region_plate</a>;</p></li>
<li><p>all random variables must be either global or Markov and sampled at every time step, so e.g. time-windowed random variables are not supported;</p></li>
<li><p>models must support broadcasting and vectorization of time <code class="docutils literal notranslate"><span class="pre">t</span></code>.</p></li>
</ul>
<p>These restrictions allow inference algorithms to vectorize over the time dimension, leading to inference algorithms with per-iteration parallel complexity sublinear in length of the time axis. The restriction on distributions allows inference algorithms to approximate parts of the model as Gaussian via moment matching, further speeding up inference. Finally, because real data is so often overdispersed relative to Binomial idealizations, the three distribution helpers provide an
<a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.distributions.binomial_dist">overdispersion</a> parameter calibrated so that in the large-population limit all distribution helpers converge to log-normal.</p>
<p>Black box inference algorithms currently include: <a class="reference external" href="http://docs.pyro.ai/en/latest/inference_algos.html#pyro.infer.svi.SVI">SVI</a> with a moment-matching approximation, and <a class="reference external" href="http://docs.pyro.ai/en/latest/mcmc.html#pyro.infer.mcmc.NUTS">NUTS</a> either with a moment-matched approximation or with an exact auxiliary variable method detailed in the <a class="reference external" href="http://pyro.ai/examples/sir_hmc.html">SIR HMC tutorial</a>. All three algorithms initialize using
<a class="reference external" href="http://docs.pyro.ai/en/latest/inference_algos.html#pyro.infer.smcfilter.SMCFilter">SMC</a> and reparameterize time dependent variables using a fast <a class="reference external" href="http://docs.pyro.ai/en/latest/infer.reparam.html#pyro.infer.reparam.haar.HaarReparam">Haar wavelet</a> transform. Default inference parameters are set for cheap approximate results; accurate results will require more steps and ideally comparison among different inference algorithms. We recommend that, when running MCMC inference, you use multiple
chains, thus making it easier to diagnose mixing issues.</p>
<p>While MCMC inference can be more accurate for a given model, SVI is much faster and thus allows richer model structure (e.g. incorporating neural networks) and more rapid <a class="reference external" href="https://www.annualreviews.org/doi/abs/10.1146/annurev-statistics-022513-115657?journalCode=statistics">model iteration</a>. We recommend starting model exploration using mean field SVI (via <code class="docutils literal notranslate"><span class="pre">.fit_svi(guide_rank=0)</span></code>), then optionally increasing accuracy using a low-rank multivariate normal guide (via
<code class="docutils literal notranslate"><span class="pre">.fit_svi(guide_rank=None)</span></code>). For even more accurate posteriors you could then try moment-matched MCMC (via <code class="docutils literal notranslate"><span class="pre">.fit_mcmc(num_quant_bins=1)</span></code>), or the most accurate and most expensive enumerated MCMC (via <code class="docutils literal notranslate"><span class="pre">.fit_mcmc(num_quant_bins=4)</span></code>). We recommend that, when fitting models with neural networks, you train via <code class="docutils literal notranslate"><span class="pre">.fit_svi()</span></code>, then freeze the network (say by omitting a <code class="docutils literal notranslate"><span class="pre">pyro.module()</span></code> statement) before optionally running MCMC inference.</p>
<section id="Modeling">
<h4>Modeling<a class="headerlink" href="#Modeling" title="Permalink to this headline">¶</a></h4>
<p>The <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#module-pyro.contrib.epidemiology.models">pyro.contrib.epidemiology.models</a> module provides a number of example models. While in principle these are reusable, we recommend forking and modifying these models for your task. Let’s take a look at one of the simplest examples, <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.models.SimpleSIRModel">SimpleSIRModel</a>. This model derives from the
<a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel">CompartmentalModel</a> base class and overrides the three standard methods using familiar Pyro modeling code in each method.</p>
<ul class="simple">
<li><p><a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.global_model">.global_model()</a> samples global parameters and packs them into a single return value (here a tuple, but any structure is allowed). The return value is available as the <code class="docutils literal notranslate"><span class="pre">params</span></code> argument to the other two methods.</p></li>
<li><p><a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.initialize">.initialize(params)</a> samples (or deterministically sets) initial values of time series, returning a dictionary mapping time series name to initial value.</p></li>
<li><p><a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.transition">.transition(params, state, t)</a> inputs global <code class="docutils literal notranslate"><span class="pre">params</span></code>, the <code class="docutils literal notranslate"><span class="pre">state</span></code> at the previous time step, and the time index <code class="docutils literal notranslate"><span class="pre">t</span></code> (which may be a slice!). It then samples flows and updates the state dict.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">SimpleSIRModel</span><span class="p">(</span><span class="n">CompartmentalModel</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">population</span><span class="p">,</span> <span class="n">recovery_time</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="n">compartments</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">)</span>  <span class="c1"># R is implicit.</span>
        <span class="n">duration</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">compartments</span><span class="p">,</span> <span class="n">duration</span><span class="p">,</span> <span class="n">population</span><span class="p">)</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">recovery_time</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">recovery_time</span> <span class="o">&gt;</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">recovery_time</span> <span class="o">=</span> <span class="n">recovery_time</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>

    <span class="k">def</span> <span class="nf">global_model</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">tau</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">recovery_time</span>
        <span class="n">R0</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;R0&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">LogNormal</span><span class="p">(</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">))</span>
        <span class="n">rho</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;rho&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">Beta</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">R0</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">rho</span>

    <span class="k">def</span> <span class="nf">initialize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
        <span class="c1"># Start with a single infection.</span>
        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;S&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
        <span class="n">R0</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">rho</span> <span class="o">=</span> <span class="n">params</span>

        <span class="c1"># Sample flows between compartments.</span>
        <span class="n">S2I</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;S2I_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                          <span class="n">infection_dist</span><span class="p">(</span><span class="n">individual_rate</span><span class="o">=</span><span class="n">R0</span> <span class="o">/</span> <span class="n">tau</span><span class="p">,</span>
                                         <span class="n">num_susceptible</span><span class="o">=</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">],</span>
                                         <span class="n">num_infectious</span><span class="o">=</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">],</span>
                                         <span class="n">population</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">))</span>
        <span class="n">I2R</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;I2R_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                          <span class="n">binomial_dist</span><span class="p">(</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">],</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">tau</span><span class="p">))</span>

        <span class="c1"># Update compartments with flows.</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">S2I</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">S2I</span> <span class="o">-</span> <span class="n">I2R</span>

        <span class="c1"># Condition on observations.</span>
        <span class="n">t_is_observed</span> <span class="o">=</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="nb">slice</span><span class="p">)</span> <span class="ow">or</span> <span class="n">t</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">duration</span>
        <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;obs_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                    <span class="n">binomial_dist</span><span class="p">(</span><span class="n">S2I</span><span class="p">,</span> <span class="n">rho</span><span class="p">),</span>
                    <span class="n">obs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">t</span><span class="p">]</span> <span class="k">if</span> <span class="n">t_is_observed</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Note that we’ve stored data in the model. These models have a scikit-learn like interface: we instantiate a model class with data, then call a <code class="docutils literal notranslate"><span class="pre">.fit_*()</span></code> method to train, then call <code class="docutils literal notranslate"><span class="pre">.predict()</span></code> on a trained model.</p>
<p>Note also that we’ve taken special care so that <code class="docutils literal notranslate"><span class="pre">t</span></code> can be either an integer or a <code class="docutils literal notranslate"><span class="pre">slice</span></code>. Under the hood, <code class="docutils literal notranslate"><span class="pre">t</span></code> is an integer during SMC initialization, a <code class="docutils literal notranslate"><span class="pre">slice</span></code> during SVI or MCMC inference, and an integer again during prediction.</p>
</section>
<section id="Generating-data">
<h4>Generating data<a class="headerlink" href="#Generating-data" title="Permalink to this headline">¶</a></h4>
<p>To check that our model generates plausible data, we can create a model with empty data and call the model’s <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.generate">.generate()</a> method. This method first calls, <code class="docutils literal notranslate"><span class="pre">.global_model()</span></code>, then calls <code class="docutils literal notranslate"><span class="pre">.initialize()</span></code>, then calls <code class="docutils literal notranslate"><span class="pre">.transition()</span></code> once per time step (based on the length of our empty data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">population</span> <span class="o">=</span> <span class="mi">10000</span>
<span class="n">recovery_time</span> <span class="o">=</span> <span class="mf">10.</span>
<span class="n">empty_data</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="mi">90</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">SimpleSIRModel</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="n">recovery_time</span><span class="p">,</span> <span class="n">empty_data</span><span class="p">)</span>

<span class="c1"># We&#39;ll repeatedly generate data until a desired number of infections is found.</span>
<span class="n">pyro</span><span class="o">.</span><span class="n">set_rng_seed</span><span class="p">(</span><span class="mi">20200709</span><span class="p">)</span>
<span class="k">for</span> <span class="n">attempt</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">):</span>
    <span class="n">synth_data</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">generate</span><span class="p">({</span><span class="s2">&quot;R0&quot;</span><span class="p">:</span> <span class="mf">2.0</span><span class="p">})</span>
    <span class="n">total_infections</span> <span class="o">=</span> <span class="n">synth_data</span><span class="p">[</span><span class="s2">&quot;S2I&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
    <span class="k">if</span> <span class="mi">4000</span> <span class="o">&lt;=</span> <span class="n">total_infections</span> <span class="o">&lt;=</span> <span class="mi">6000</span><span class="p">:</span>
        <span class="k">break</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Simulated </span><span class="si">{}</span><span class="s2"> infections after </span><span class="si">{}</span><span class="s2"> attempts&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">total_infections</span><span class="p">,</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">attempt</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Simulated 4055.0 infections after 6 attempts
</pre></div></div>
</div>
<p>The generated data contains both global variables and time series, packed into tensors.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">synth_data</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.shape = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">value</span><span class="o">.</span><span class="n">shape</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
I.shape = (90,)
I2R.shape = (90,)
R0.shape = ()
S.shape = (90,)
S2I.shape = (90,)
obs.shape = (90,)
rho.shape = ()
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">synth_data</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="k">if</span> <span class="n">value</span><span class="o">.</span><span class="n">dim</span><span class="p">():</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">empty_data</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;time step&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;individuals&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;best&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Synthetic time series&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/epi_intro_10_0.png" src="_images/epi_intro_10_0.png" />
</div>
</div>
</section>
<section id="Inference">
<h4>Inference<a class="headerlink" href="#Inference" title="Permalink to this headline">¶</a></h4>
<p>Next let’s recover estimates of the latent variables given only observations <code class="docutils literal notranslate"><span class="pre">obs</span></code>. To do this we’ll create a new model instance from the synthetic observations.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">obs</span> <span class="o">=</span> <span class="n">synth_data</span><span class="p">[</span><span class="s2">&quot;obs&quot;</span><span class="p">]</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">SimpleSIRModel</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="n">recovery_time</span><span class="p">,</span> <span class="n">obs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">CompartmentalModel</span></code> provides a number of inference algorithms. The cheapest and most scalable algorithm is SVI, avilable via the <code class="docutils literal notranslate"><span class="pre">.fit_svi()</span></code> method. This method returns a list of losses to help us diagnose convergence; the fitted parameters are stored in the model object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="n">losses</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit_svi</span><span class="p">(</span><span class="n">num_steps</span><span class="o">=</span><span class="mi">101</span> <span class="k">if</span> <span class="n">smoke_test</span> <span class="k">else</span> <span class="mi">2001</span><span class="p">,</span>
                       <span class="n">jit</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO     Heuristic init: R0=1.83, rho=0.546
INFO     Running inference...
INFO     step 0 loss = 6.808
INFO     step 200 loss = 9.099
INFO     step 400 loss = 7.384
INFO     step 600 loss = 4.401
INFO     step 800 loss = 3.428
INFO     step 1000 loss = 3.242
INFO     step 1200 loss = 3.13
INFO     step 1400 loss = 3.016
INFO     step 1600 loss = 3.029
INFO     step 1800 loss = 3.05
INFO     step 2000 loss = 3.017
INFO     SVI took 12.7 seconds, 157.9 step/sec
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 12.8 s, sys: 278 ms, total: 13.1 s
Wall time: 13.2 s
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">losses</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;SVI step&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;loss&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">losses</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">losses</span><span class="p">[</span><span class="mi">50</span><span class="p">:]));</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/epi_intro_15_0.png" src="_images/epi_intro_15_0.png" />
</div>
</div>
<p>After inference, samples of latent variables are stored in the <code class="docutils literal notranslate"><span class="pre">.samples</span></code> attribute. These are primarily for internal use, and do not contain the full set of latent variables.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.shape = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">value</span><span class="o">.</span><span class="n">shape</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
R0.shape = (100, 1)
auxiliary.shape = (100, 1, 2, 90)
rho.shape = (100, 1)
</pre></div></div>
</div>
</section>
<section id="Prediction">
<h4>Prediction<a class="headerlink" href="#Prediction" title="Permalink to this headline">¶</a></h4>
<p>After inference we can both examine latent variables and forecast forward using the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.predict">.predict()</a> method. First let’s simply predict latent variables.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="n">samples</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO     Predicting latent variables for 90 time steps...
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 113 ms, sys: 2.82 ms, total: 116 ms
Wall time: 115 ms
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.shape = </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">value</span><span class="o">.</span><span class="n">shape</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
I.shape = (100, 90)
I2R.shape = (100, 90)
R0.shape = (100, 1)
S.shape = (100, 90)
S2I.shape = (100, 90)
auxiliary.shape = (100, 1, 2, 90)
obs.shape = (100, 90)
rho.shape = (100, 1)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;R0&quot;</span><span class="p">,</span> <span class="s2">&quot;rho&quot;</span><span class="p">]</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Posterior estimates of global parameters&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">axes</span><span class="p">,</span> <span class="n">names</span><span class="p">):</span>
    <span class="n">truth</span> <span class="o">=</span> <span class="n">synth_data</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">name</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;posterior&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">truth</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;truth&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(())</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;best&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/epi_intro_21_0.png" src="_images/epi_intro_21_0.png" />
</div>
</div>
<p>Notice that while the inference recovers the basic reproductive number <code class="docutils literal notranslate"><span class="pre">R0</span></code>, it poorly estimates the response rate <code class="docutils literal notranslate"><span class="pre">rho</span></code> and underestimates its uncertainty. While perfect inference would provide better uncertainty estimates, the response rate is known to be difficult to recover from data. Ideally the model can either incorporate a narrower prior, either obtained by testing a random sample of the population, or by more accurate observations, e.g. counting deaths rather than confirmed
infections.</p>
</section>
<section id="Forecasting">
<h4>Forecasting<a class="headerlink" href="#Forecasting" title="Permalink to this headline">¶</a></h4>
<p>We can forecast forward by passing a <code class="docutils literal notranslate"><span class="pre">forecast</span></code> argument to the <code class="docutils literal notranslate"><span class="pre">.predict()</span></code> method, specifying the number of time steps ahead we’d like to forecast. The returned <code class="docutils literal notranslate"><span class="pre">sample</span></code> will contain time values during both the first observed time interval (here 90 days) and the forecasted window (say 30 days).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">time</span>
<span class="n">samples</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">forecast</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO     Predicting latent variables for 90 time steps...
INFO     Forecasting 30 steps ahead...
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 2 µs, sys: 1 µs, total: 3 µs
Wall time: 5.96 µs
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_forecast</span><span class="p">(</span><span class="n">samples</span><span class="p">):</span>
    <span class="n">duration</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">empty_data</span><span class="p">)</span>
    <span class="n">forecast</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">-</span> <span class="n">duration</span>
    <span class="n">num_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="s2">&quot;R0&quot;</span><span class="p">])</span>

    <span class="n">time</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">duration</span> <span class="o">+</span> <span class="n">forecast</span><span class="p">)</span>
    <span class="n">S2I</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="s2">&quot;S2I&quot;</span><span class="p">]</span>
    <span class="n">median</span> <span class="o">=</span> <span class="n">S2I</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">p05</span> <span class="o">=</span> <span class="n">S2I</span><span class="o">.</span><span class="n">kthvalue</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="mf">0.5</span> <span class="o">+</span> <span class="mf">0.05</span> <span class="o">*</span> <span class="n">num_samples</span><span class="p">)),</span> <span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">p95</span> <span class="o">=</span> <span class="n">S2I</span><span class="o">.</span><span class="n">kthvalue</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="mf">0.5</span> <span class="o">+</span> <span class="mf">0.95</span> <span class="o">*</span> <span class="n">num_samples</span><span class="p">)),</span> <span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">time</span><span class="p">,</span> <span class="n">p05</span><span class="p">,</span> <span class="n">p95</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;90% CI&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time</span><span class="p">,</span> <span class="n">median</span><span class="p">,</span> <span class="s2">&quot;r-&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;median&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time</span><span class="p">[:</span><span class="n">duration</span><span class="p">],</span> <span class="n">obs</span><span class="p">,</span> <span class="s2">&quot;k.&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;observed&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time</span><span class="p">[:</span><span class="n">duration</span><span class="p">],</span> <span class="n">synth_data</span><span class="p">[</span><span class="s2">&quot;S2I&quot;</span><span class="p">],</span> <span class="s2">&quot;k--&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;truth&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">duration</span> <span class="o">-</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">time</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;day after first infection&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;new infections per day&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;New infections in population of </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">population</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

<span class="n">plot_forecast</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/epi_intro_25_0.png" src="_images/epi_intro_25_0.png" />
</div>
</div>
<p>It looks like the mean field guide underestimates uncertainty. To improve uncertainty estimates we can instead try MCMC inference. In this simple model MCMC is only a small factor slower than SVI; in more complex models MCMC can be multiple orders of magnitude slower than SVI.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">SimpleSIRModel</span><span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="n">recovery_time</span><span class="p">,</span> <span class="n">obs</span><span class="p">)</span>
<span class="n">mcmc</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit_mcmc</span><span class="p">(</span><span class="n">num_samples</span><span class="o">=</span><span class="mi">4</span> <span class="k">if</span> <span class="n">smoke_test</span> <span class="k">else</span> <span class="mi">400</span><span class="p">,</span>
                      <span class="n">jit_compile</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO     Running inference...
Warmup:   0%|          | 0/800 [00:00, ?it/s]INFO        Heuristic init: R0=2.05, rho=0.437
Sample: 100%|██████████| 800/800 [01:44,  7.63it/s, step size=1.07e-01, acc. prob=0.890]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPU times: user 1min 43s, sys: 1.2 s, total: 1min 44s
Wall time: 1min 44s
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">forecast</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">plot_forecast</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO     Predicting latent variables for 90 time steps...
INFO     Forecasting 30 steps ahead...
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/epi_intro_28_1.png" src="_images/epi_intro_28_1.png" />
</div>
</div>
</section>
</section>
<section id="Advanced-modeling">
<h3>Advanced modeling<a class="headerlink" href="#Advanced-modeling" title="Permalink to this headline">¶</a></h3>
<p>So far we’ve seen how to create a simple univariate model, fit the model to data, and predict and forecast future data. Next let’s consider more advanced modeling techniques:</p>
<ul class="simple">
<li><p><a class="reference external" href="#Regional-models">regional models</a> that couple compartments among multiple aggregated regions;</p></li>
<li><p><a class="reference external" href="#Phylogenetic-likelihoods">phylogenetic likelihoods</a> to incorporate genetic sequencing data;</p></li>
<li><p><a class="reference external" href="#Heterogeneous-models">heterogeneous models</a> with time-varying latent variables; and</p></li>
<li><p><a class="reference external" href="#Complex-compartment-flow">Complex compartment flow</a> for models with non-linear transitions.</p></li>
</ul>
<section id="Regional-models">
<h4>Regional models<a class="headerlink" href="#Regional-models" title="Permalink to this headline">¶</a></h4>
<p>Epidemiology models vary in their level of detail. At the coarse-grained extreme are univariate aggregate models as we saw above. At the fine-grained extreme are network models where each individual’s state is tracked and infections occur along edges of a sparse graph (<code class="docutils literal notranslate"><span class="pre">pyro.contrib.epidemiology</span></code> does not implement network models). We now consider an mid-level model where each of many regions (e.g. countries or zip codes) is tracked in aggregate, and infections occur both within regions and
between pairs of regions. In Pyro we model multiple regions with a <a class="reference external" href="http://docs.pyro.ai/en/stable/primitives.html#pyro.primitives.plate">plate</a>. Pyro’s <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel">CompartmentalModel</a> class does not support general <code class="docutils literal notranslate"><span class="pre">pyro.plate</span></code> syntax, but it does support a single special <code class="docutils literal notranslate"><span class="pre">self.region_plate</span></code> for regional models. This plate is available iff a <code class="docutils literal notranslate"><span class="pre">CompartmentalModel</span></code> is initialized
with a vector <code class="docutils literal notranslate"><span class="pre">population</span></code>, and the size of the <code class="docutils literal notranslate"><span class="pre">region_plate</span></code> will be the length of the <code class="docutils literal notranslate"><span class="pre">population</span></code> vector.</p>
<p>Let’s take a look at the example <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.models.RegionalSIRModel">RegionalSIRModel</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">RegionalSIRModel</span><span class="p">(</span><span class="n">CompartmentalModel</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">population</span><span class="p">,</span> <span class="n">coupling</span><span class="p">,</span> <span class="n">recovery_time</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="n">duration</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="n">num_regions</span><span class="p">,</span> <span class="o">=</span> <span class="n">population</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">assert</span> <span class="n">coupling</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">num_regions</span><span class="p">,</span> <span class="n">num_regions</span><span class="p">)</span>
        <span class="k">assert</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">coupling</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">assert</span> <span class="p">(</span><span class="n">coupling</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">recovery_time</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">recovery_time</span> <span class="o">&gt;</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
            <span class="c1"># Data tensors should be oriented as (time, region).</span>
            <span class="k">assert</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="n">duration</span><span class="p">,</span> <span class="n">num_regions</span><span class="p">)</span>
        <span class="n">compartments</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">)</span>  <span class="c1"># R is implicit.</span>

        <span class="c1"># We create a regional model by passing a vector of populations.</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">compartments</span><span class="p">,</span> <span class="n">duration</span><span class="p">,</span> <span class="n">population</span><span class="p">,</span> <span class="n">approximate</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;I&quot;</span><span class="p">,))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">coupling</span> <span class="o">=</span> <span class="n">coupling</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">recovery_time</span> <span class="o">=</span> <span class="n">recovery_time</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>

    <span class="k">def</span> <span class="nf">global_model</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Assume recovery time is a known constant.</span>
        <span class="n">tau</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">recovery_time</span>

        <span class="c1"># Assume reproductive number is unknown but homogeneous.</span>
        <span class="n">R0</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;R0&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">LogNormal</span><span class="p">(</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">))</span>

        <span class="c1"># Assume response rate is heterogeneous and model it with a</span>
        <span class="c1"># hierarchical Gamma-Beta prior.</span>
        <span class="n">rho_c1</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;rho_c1&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">Gamma</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">rho_c0</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;rho_c0&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">Gamma</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
        <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">region_plate</span><span class="p">:</span>
            <span class="n">rho</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;rho&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">Beta</span><span class="p">(</span><span class="n">rho_c1</span><span class="p">,</span> <span class="n">rho_c0</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">R0</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">rho</span>

    <span class="k">def</span> <span class="nf">initialize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
        <span class="c1"># Start with a single infection in region 0.</span>
        <span class="n">I</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>
        <span class="n">I</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">S</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">-</span> <span class="n">I</span>
        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;S&quot;</span><span class="p">:</span> <span class="n">S</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">:</span> <span class="n">I</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
        <span class="n">R0</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">rho</span> <span class="o">=</span> <span class="n">params</span>

        <span class="c1"># Account for infections from all regions. This uses approximate (point</span>
        <span class="c1"># estimate) counts I_approx for infection from other regions, but uses</span>
        <span class="c1"># the exact (enumerated) count I for infections from one&#39;s own region.</span>
        <span class="n">I_coupled</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I_approx&quot;</span><span class="p">]</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">coupling</span>
        <span class="n">I_coupled</span> <span class="o">=</span> <span class="n">I_coupled</span> <span class="o">+</span> <span class="p">(</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I_approx&quot;</span><span class="p">])</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">coupling</span><span class="o">.</span><span class="n">diag</span><span class="p">()</span>
        <span class="n">I_coupled</span> <span class="o">=</span> <span class="n">I_coupled</span><span class="o">.</span><span class="n">clamp</span><span class="p">(</span><span class="nb">min</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># In case I_approx is negative.</span>
        <span class="n">pop_coupled</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">@</span> <span class="bp">self</span><span class="o">.</span><span class="n">coupling</span>

        <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">region_plate</span><span class="p">:</span>
            <span class="c1"># Sample flows between compartments.</span>
            <span class="n">S2I</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;S2I_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                              <span class="n">infection_dist</span><span class="p">(</span><span class="n">individual_rate</span><span class="o">=</span><span class="n">R0</span> <span class="o">/</span> <span class="n">tau</span><span class="p">,</span>
                                             <span class="n">num_susceptible</span><span class="o">=</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">],</span>
                                             <span class="n">num_infectious</span><span class="o">=</span><span class="n">I_coupled</span><span class="p">,</span>
                                             <span class="n">population</span><span class="o">=</span><span class="n">pop_coupled</span><span class="p">))</span>
            <span class="n">I2R</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;I2R_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                              <span class="n">binomial_dist</span><span class="p">(</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">],</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">tau</span><span class="p">))</span>

            <span class="c1"># Update compartments with flows.</span>
            <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">S2I</span>
            <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">S2I</span> <span class="o">-</span> <span class="n">I2R</span>

            <span class="c1"># Condition on observations.</span>
            <span class="n">t_is_observed</span> <span class="o">=</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="nb">slice</span><span class="p">)</span> <span class="ow">or</span> <span class="n">t</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">duration</span>
            <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;obs_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                        <span class="n">binomial_dist</span><span class="p">(</span><span class="n">S2I</span><span class="p">,</span> <span class="n">rho</span><span class="p">),</span>
                        <span class="n">obs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">t</span><span class="p">]</span> <span class="k">if</span> <span class="n">t_is_observed</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The main differences from the earlier univariate model are that: we assume <code class="docutils literal notranslate"><span class="pre">population</span></code> is a vector of length <code class="docutils literal notranslate"><span class="pre">num_regions</span></code>, we sample all compartmental variables and some global variables inside the <code class="docutils literal notranslate"><span class="pre">region_plate</span></code>, and we compute coupled vectors <code class="docutils literal notranslate"><span class="pre">I_coupled</span></code> and <code class="docutils literal notranslate"><span class="pre">pop_coupled</span></code> of the effective number of infected individuals and population accounting for both intra-region and inter-region infections. Among global variables we have chosen for demonstration purposes to make <code class="docutils literal notranslate"><span class="pre">tau</span></code> a fixed
single number, <code class="docutils literal notranslate"><span class="pre">R0</span></code> a single latent variable shared among all regions, and <code class="docutils literal notranslate"><span class="pre">rho</span></code> a local latent variable that can take a different value for each region. Note that while <code class="docutils literal notranslate"><span class="pre">rho</span></code> is not shared among regions, we have created a hierarchical model whereby <code class="docutils literal notranslate"><span class="pre">rho</span></code>’s parent variables are shared among regions. While some of our variables are region-global and some region-local, only the compartmental variables are both region-local and time-dependent; all other parameters are fixed for all time.
See the <a class="reference external" href="#Heterogeneous-models">heterogeneous models</a> section below for time-dependent latent variables.</p>
<p>Note that Pyro’s enumerated MCMC strategy (<code class="docutils literal notranslate"><span class="pre">.fit_mcmc()</span></code> with <code class="docutils literal notranslate"><span class="pre">num_quant_bins</span> <span class="pre">&gt;</span> <span class="pre">1</span></code>) requires extra logic to use a mean-field approximation across compartments: we pass <code class="docutils literal notranslate"><span class="pre">approximate=(&quot;I&quot;,)</span></code> to the constructor and force compartements to iteract via <code class="docutils literal notranslate"><span class="pre">state[&quot;I_approx&quot;]</span></code> rather than <code class="docutils literal notranslate"><span class="pre">state[&quot;I&quot;]</span></code>. This code is not required for SVI inference or for moment-matched MCMC inference (<code class="docutils literal notranslate"><span class="pre">.fit_mcmc()</span></code> with the default <code class="docutils literal notranslate"><span class="pre">num_quant_bins=0</span></code>).</p>
<p>See the <a class="reference external" href="http://pyro.ai/examples/epi_regional.html">Epidemiology: regional models</a> example for a demonstration of how to generate data, train, predict, and forecast with regional models.</p>
</section>
<section id="Phylogenetic-likelihoods">
<h4>Phylogenetic likelihoods<a class="headerlink" href="#Phylogenetic-likelihoods" title="Permalink to this headline">¶</a></h4>
<p>Epidemiological parameters can be difficult to identify from aggregate observations alone. However some parameters like the superspreading parameter <code class="docutils literal notranslate"><span class="pre">k</span></code> can be more accurately identified by combining aggregate count data with viral phylogenetic trees reconstructed from viral genetic sequencing data <a class="reference external" href="#1">(Li et al. 2017)</a>. Pyro implements a <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.distributions.CoalescentRateLikelihood">CoalescentRateLikelihood</a> class to compute a
population likelihood <code class="docutils literal notranslate"><span class="pre">p(I|phylogeny)</span></code> given statistics of a phylogenetic tree (or a batch of tree samples). The statistics needed are exactly the times of each sampling event (i.e. when a viral genome was sequenced) and the times of genetic coalescent events in a binary phylogenetic tree; let us call these two vectors <code class="docutils literal notranslate"><span class="pre">leaf_times</span></code> and <code class="docutils literal notranslate"><span class="pre">coal_times</span></code>, respectively, where <code class="docutils literal notranslate"><span class="pre">len(leaf_times)</span> <span class="pre">==</span> <span class="pre">1</span> <span class="pre">+</span> <span class="pre">len(coal_times)</span></code> for binary trees. Pyro provides a helper
<a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.distributions.coalescent.bio_phylo_to_times">bio_phylo_to_times()</a> to extract these statistics from a <a class="reference external" href="https://biopython.readthedocs.io/en/latest/api/Bio.Phylo.BaseTree.html#Bio.Phylo.BaseTree.Clade">Bio.Phylo tree objects</a>; in turn <a class="reference external" href="https://biopython-tutorial.readthedocs.io/en/latest/notebooks/13%20-%20Phylogenetics%20with%20Bio.Phylo.html">Bio.Phylo</a> can parse many file formats of phylogenetic trees.</p>
<p>Let’s take a look at the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.models.SuperspreadingSEIRModel">SuperspreadingSEIRModel</a> which includes a phylogenetic likelihood. We’ll focus on the phylogenetic parts of the model:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">SuperspreadingSEIRModel</span><span class="p">(</span><span class="n">CompartmentalModel</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">population</span><span class="p">,</span> <span class="n">incubation_time</span><span class="p">,</span> <span class="n">recovery_time</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
                 <span class="n">leaf_times</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">coal_times</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">compartments</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">)</span>  <span class="c1"># R is implicit.</span>
        <span class="n">duration</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">compartments</span><span class="p">,</span> <span class="n">duration</span><span class="p">,</span> <span class="n">population</span><span class="p">)</span>
        <span class="o">...</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">coal_likelihood</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">CoalescentRateLikelihood</span><span class="p">(</span>
            <span class="n">leaf_times</span><span class="p">,</span> <span class="n">coal_times</span><span class="p">,</span> <span class="n">duration</span><span class="p">)</span>
    <span class="o">...</span>

    <span class="k">def</span> <span class="nf">transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
        <span class="o">...</span>
        <span class="c1"># Condition on observations.</span>
        <span class="n">t_is_observed</span> <span class="o">=</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="nb">slice</span><span class="p">)</span> <span class="ow">or</span> <span class="n">t</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">duration</span>
        <span class="n">R</span> <span class="o">=</span> <span class="n">R0</span> <span class="o">*</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span>
        <span class="n">coal_rate</span> <span class="o">=</span> <span class="n">R</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.</span> <span class="o">+</span> <span class="mf">1.</span> <span class="o">/</span> <span class="n">k</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">tau_i</span> <span class="o">*</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mf">1e-8</span><span class="p">)</span>
        <span class="n">pyro</span><span class="o">.</span><span class="n">factor</span><span class="p">(</span><span class="s2">&quot;coalescent_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">coal_likelihood</span><span class="p">(</span><span class="n">coal_rate</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">t_is_observed</span> <span class="k">else</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="mf">0.</span><span class="p">))</span>
</pre></div>
</div>
<p>We first constructed a <code class="docutils literal notranslate"><span class="pre">CoalescentRateLikelihood</span></code> object to be used throughout inference and prediction; this performs preprocessing work once so that it is cheap to evaluate <code class="docutils literal notranslate"><span class="pre">self.coal_likelihood(...)</span></code>. Note that <code class="docutils literal notranslate"><span class="pre">(leaf_times,</span> <span class="pre">coal_times)</span></code> should be in units of time steps, the same time steps as the time index <code class="docutils literal notranslate"><span class="pre">t</span></code> and <code class="docutils literal notranslate"><span class="pre">duration</span></code>. Typically <code class="docutils literal notranslate"><span class="pre">leaf_times</span></code> are in <code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">duration)</span></code>, but <code class="docutils literal notranslate"><span class="pre">coal_times</span></code> precede <code class="docutils literal notranslate"><span class="pre">leaf_times</span></code> (as points of common ancestry), and may be negative. The
likelihood involves the coalescent rate <code class="docutils literal notranslate"><span class="pre">coal_rate</span></code> in a coalescent process; we can compute this from an epidemiological model. In this superspreading model <code class="docutils literal notranslate"><span class="pre">coal_rate</span></code> depends on the reproductive number <code class="docutils literal notranslate"><span class="pre">R</span></code>, the superspreading parameter <code class="docutils literal notranslate"><span class="pre">k</span></code>, the incubation time <code class="docutils literal notranslate"><span class="pre">tau_i</span></code>, and the current number of infected individuals <code class="docutils literal notranslate"><span class="pre">state[&quot;I&quot;]</span></code> <a class="reference external" href="#1">(Li et al. 2017)</a>.</p>
</section>
<section id="Heterogeneous-models">
<h4>Heterogeneous models<a class="headerlink" href="#Heterogeneous-models" title="Permalink to this headline">¶</a></h4>
<p>Epidemiological parameters often vary in time, due to human interventions, changes in weather, and other external factors. We can model real-valued time-varying latent variables in <code class="docutils literal notranslate"><span class="pre">CompartmentalModel</span></code> by moving static latent variables from <code class="docutils literal notranslate"><span class="pre">.global_model()</span></code> to <code class="docutils literal notranslate"><span class="pre">.initialize()</span></code> and <code class="docutils literal notranslate"><span class="pre">.transition()</span></code>. For example we can model a reproductive number under Brownian drift in log-space by initializing at a random <code class="docutils literal notranslate"><span class="pre">R0</span></code> and multiplying by a drifting factor, as in the
<a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.models.HeterogeneousSIRModel">HeterogeneousSIRModel</a> example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">HeterogeneousSIRModel</span><span class="p">(</span><span class="n">CompartmentalModel</span><span class="p">):</span>
    <span class="o">...</span>
    <span class="k">def</span> <span class="nf">global_model</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">tau</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">recovery_time</span>
        <span class="n">R0</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;R0&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">LogNormal</span><span class="p">(</span><span class="mf">0.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">))</span>
        <span class="n">rho</span> <span class="o">=</span> <span class="o">...</span>
        <span class="k">return</span> <span class="n">R0</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">rho</span>

    <span class="k">def</span> <span class="nf">initialize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
        <span class="c1"># Start with a single infection.</span>
        <span class="c1"># We also store the initial beta value in the state dict.</span>
        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;S&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">population</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">:</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="mf">1.</span><span class="p">)}</span>

    <span class="k">def</span> <span class="nf">transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
        <span class="n">R0</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">rho</span> <span class="o">=</span> <span class="n">params</span>
        <span class="c1"># Sample heterogeneous variables.</span>
        <span class="c1"># This assumes beta slowly drifts via Brownian motion in log space.</span>
        <span class="n">beta</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;beta_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                           <span class="n">dist</span><span class="o">.</span><span class="n">LogNormal</span><span class="p">(</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;beta&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">log</span><span class="p">(),</span> <span class="mf">0.1</span><span class="p">))</span>
        <span class="n">Rt</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">deterministic</span><span class="p">(</span><span class="s2">&quot;Rt_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span> <span class="n">R0</span> <span class="o">*</span> <span class="n">beta</span><span class="p">)</span>

        <span class="c1"># Sample flows between compartments.</span>
        <span class="n">S2I</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;S2I_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span>
                          <span class="n">infection_dist</span><span class="p">(</span><span class="n">individual_rate</span><span class="o">=</span><span class="n">Rt</span> <span class="o">/</span> <span class="n">tau</span><span class="p">,</span>
                                         <span class="n">num_susceptible</span><span class="o">=</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">],</span>
                                         <span class="n">num_infectious</span><span class="o">=</span><span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">],</span>
                                         <span class="n">population</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">population</span><span class="p">))</span>
        <span class="o">...</span>
        <span class="c1"># Update compartments and heterogeneous variables.</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">S2I</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">S2I</span> <span class="o">-</span> <span class="n">I2R</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;beta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">beta</span>  <span class="c1"># We store the latest beta value in the state dict.</span>
        <span class="o">...</span>
</pre></div>
</div>
<p>Here we deterministically initialize a scale factor <code class="docutils literal notranslate"><span class="pre">beta</span> <span class="pre">=</span> <span class="pre">1</span></code> in <code class="docutils literal notranslate"><span class="pre">.initialize()</span></code> then let it drift via log-Brownian motion. We also need to update <code class="docutils literal notranslate"><span class="pre">state[&quot;beta&quot;]</span></code> just as we update the compartmental variables. Now <code class="docutils literal notranslate"><span class="pre">beta</span></code> will be provided as a time series when we <code class="docutils literal notranslate"><span class="pre">.predict()</span></code>. While we could have written <code class="docutils literal notranslate"><span class="pre">Rt</span> <span class="pre">=</span> <span class="pre">R0</span> <span class="pre">*</span> <span class="pre">beta</span></code>, we instead wrapped this computation in a <code class="docutils literal notranslate"><span class="pre">pyro.deterministic</span></code> thereby exposing <code class="docutils literal notranslate"><span class="pre">Rt</span></code> as another time series provided by <code class="docutils literal notranslate"><span class="pre">.predict()</span></code>. Note that we could have
instead sampled <code class="docutils literal notranslate"><span class="pre">R0</span></code> in <code class="docutils literal notranslate"><span class="pre">.initialize()</span></code> and let <code class="docutils literal notranslate"><span class="pre">Rt</span></code> drift directly, rather than introducing a scale factor <code class="docutils literal notranslate"><span class="pre">beta</span></code>. However separating the two into a non-centered form improves geometry <a class="reference external" href="#2">(Betancourt and Girolami 2013)</a>.</p>
<p>It is also easy to pass in time-varying covariates as tensors, in the same way we have passed in <code class="docutils literal notranslate"><span class="pre">data</span></code> to the constructors of all example models. To predict the effects of different causal interventions, you can pass in a covariate that is longer than <code class="docutils literal notranslate"><span class="pre">duration</span></code>, run inference (looking only at the first <code class="docutils literal notranslate"><span class="pre">[0,duration)</span></code> entries), then mutate entries of the covariate after <code class="docutils literal notranslate"><span class="pre">duration</span></code> and generate different <code class="docutils literal notranslate"><span class="pre">.predict()</span></code>ions.</p>
</section>
<section id="Complex-compartment-flow">
<h4>Complex compartment flow<a class="headerlink" href="#Complex-compartment-flow" title="Permalink to this headline">¶</a></h4>
<p>The <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel">CompartmentalModel</a> class assumes by default that the compartments are arranged linearly and terminate in an implicit terminal compartment named “R”, for example S-I-R, S-E-I-R or boxcar models like S-E1-E2-I1-I2-I3-R. To describe other more complex flows between compartments, you can override the
<a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.compartmental.CompartmentalModel.compute_flows">.compute_flows()</a> method. However currently there is no support for flows with undirected loops (e.g. S-I-S).</p>
<p>Let’s create a branching SIRD model with possible flows</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>S → I → R
      ↘
        D
</pre></div>
</div>
<p>As with other models, we’ll keep the “R” state implicit (although we could equally keep the “D” state implicit and the “R” state explicit). In the <code class="docutils literal notranslate"><span class="pre">.compute_flows()</span></code> method, we’ll input a pair of states and we’ll need to compute three flow values: <code class="docutils literal notranslate"><span class="pre">S2I</span></code>, <code class="docutils literal notranslate"><span class="pre">I2R</span></code>, and <code class="docutils literal notranslate"><span class="pre">I2D</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">SIRDModel</span><span class="p">(</span><span class="n">CompartmentalModel</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">population</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
        <span class="n">compartments</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="s2">&quot;I&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">)</span>
        <span class="n">duration</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">compartments</span><span class="p">,</span> <span class="n">duration</span><span class="p">,</span> <span class="n">population</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>

    <span class="k">def</span> <span class="nf">compute_flows</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prev</span><span class="p">,</span> <span class="n">curr</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
        <span class="n">S2I</span> <span class="o">=</span> <span class="n">prev</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">curr</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span>  <span class="c1"># S can only go in one direction.</span>
        <span class="n">I2D</span> <span class="o">=</span> <span class="n">curr</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">prev</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">]</span>  <span class="c1"># D can only have come from one direction.</span>
        <span class="c1"># Now by conservation at I, change + inflows + outflows = 0,</span>
        <span class="c1"># so we can solve for the single unknown I2R.</span>
        <span class="n">I2R</span> <span class="o">=</span> <span class="n">prev</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">curr</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">S2I</span> <span class="o">-</span> <span class="n">I2D</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;S2I_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">):</span> <span class="n">S2I</span><span class="p">,</span>
            <span class="s2">&quot;I2D_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">):</span> <span class="n">I2D</span><span class="p">,</span>
            <span class="s2">&quot;I2R_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">):</span> <span class="n">I2R</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="o">...</span>
    <span class="k">def</span> <span class="nf">transition</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">t</span><span class="p">):</span>
        <span class="o">...</span>
        <span class="c1"># Sample flows between compartments.</span>
        <span class="n">S2I</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;S2I_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span> <span class="o">...</span><span class="p">)</span>
        <span class="n">I2D</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;I2D_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span> <span class="o">...</span><span class="p">)</span>
        <span class="n">I2R</span> <span class="o">=</span> <span class="n">pyro</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="s2">&quot;I2R_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">t</span><span class="p">),</span> <span class="o">...</span><span class="p">)</span>

        <span class="c1"># Update compartments with flows.</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;S&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">S2I</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">S2I</span> <span class="o">-</span> <span class="n">I2D</span> <span class="o">-</span> <span class="n">I2R</span>
        <span class="n">state</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">state</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">I2D</span>
        <span class="o">...</span>
</pre></div>
</div>
<p>Note you can name the dict keys anything you want, as long as they match your sample statements in <code class="docutils literal notranslate"><span class="pre">.transition()</span></code> and you correctly reverse the flow computation in <code class="docutils literal notranslate"><span class="pre">.transition()</span></code>. During inference Pyro will check that the <code class="docutils literal notranslate"><span class="pre">.compute_flows()</span></code> and <code class="docutils literal notranslate"><span class="pre">.transition()</span></code> computations agree. Take care to avoid in-place PyTorch operations, since these can modify the tensors rather than the dictionary:</p>
<div class="highlight-diff notranslate"><div class="highlight"><pre><span></span><span class="gi">+ state[&quot;S&quot;] = state[&quot;S&quot;] - S2I  # Correct</span><span class="w"></span>
<span class="gd">- state[&quot;S&quot;] -= S2I              # AVOID: may corrupt tensors</span><span class="w"></span>
</pre></div>
</div>
<p>For a slightly more complex example, take a look at the <a class="reference external" href="http://docs.pyro.ai/en/latest/contrib.epidemiology.html#pyro.contrib.epidemiology.models.SimpleSEIRDModel">SimpleSEIRDModel</a>.</p>
</section>
</section>
<section id="References">
<h3>References<a class="headerlink" href="#References" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li><p>Lucy M. Li, Nicholas C. Grassly, Christophe Fraser (2017) “Quantifying Transmission Heterogeneity Using Both Pathogen Phylogenies and Incidence Time Series” <a class="reference external" href="https://academic.oup.com/mbe/article/34/11/2982/3952784">https://academic.oup.com/mbe/article/34/11/2982/3952784</a></p></li>
<li><ol class="upperalpha simple" start="13">
<li><ol class="upperalpha simple" start="10">
<li><p>Betancourt, Mark Girolami (2013) “Hamiltonian Monte Carlo for Hierarchical Models” <a class="reference external" href="https://arxiv.org/abs/1312.0906">https://arxiv.org/abs/1312.0906</a></p></li>
</ol>
</li>
</ol>
</li>
</ol>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="dkl.html" class="btn btn-neutral float-left" title="Example: Deep Kernel Learning" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="epi_sir.html" class="btn btn-neutral float-right" title="Example: Univariate epidemiological models" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Pyro Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>